src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10610,6 +10592,84 @@ private bool IsTargetTypedObjectCreation()
             }
         }
 
+#nullable enable
+
+        private ExpressionSyntax ParseWithExpression(ExpressionSyntax receiverExpression, SyntaxToken withKeyword)
+        {
+            var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
+
+            var assignments = _pool.AllocateSeparated<AssignmentExpressionSyntax>();
+            if (CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+            {
+tryAgain:
+                if (CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                {
+                    // First assignment
+                    assignments.Add(parseAssignment());
+
+                    // Rest
+                    while (CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+                    {
+                        if (CurrentToken.Kind == SyntaxKind.CommaToken || IsNamedAssignment())
+                        {
+                            assignments.AddSeparator(EatToken(SyntaxKind.CommaToken));
+
+                            if (CurrentToken.Kind == SyntaxKind.CloseBraceToken)
+                            {
+                                break;
+                            }
+                            assignments.Add(parseAssignment());
+                        }
+                        else if (skipBadTokens(ref openBrace, SyntaxKind.CommaToken) == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+                    }
+                }
+                else if (skipBadTokens(ref openBrace, SyntaxKind.IdentifierToken) == PostSkipAction.Continue)
```

Comment:
>skipBadTokens [](start = 25, length = 13)

I'm afraid this will result in pretty bad error recovery.  I suggest instead simply parsing an expression.  If it ends up being an assignment expression, use that.  If not, add a missing "identifier equals" with a syntax error and use that.